search/utils.py

```diff
@@ -0,0 +1,366 @@
+    """
+
+    # Create the index if it doesn't exist.
+    if not CONN.indices.exists(INDEX_NAME):
+        CONN.indices.create(INDEX_NAME)
+    # Delete the mapping if an older version exists.
+    if CONN.indices.exists_type(index=INDEX_NAME, doc_type=DOC_TYPE):
+        CONN.indices.delete_mapping(index=INDEX_NAME, doc_type=DOC_TYPE)
+
+    mapping = Mapping(DOC_TYPE)
+
+    mapping.field("course", "string", index="not_analyzed")
+    mapping.field("description_path", "string", index="no")
+    mapping.field("description", "string", index="analyzed")
+    mapping.field("preview_url", "string", index="no")
+    mapping.field("repository", "string", index="not_analyzed")
+    mapping.field("resource_type", "string", index="analyzed")
+    mapping.field("content_xml", "string", index="no")
+    mapping.field("content_stripped", "string", index="analyzed")
+    mapping.field("run", "string", index="not_analyzed")
+    mapping.field("titlesort", "string", index="no")
+    mapping.field("title", "string", index="analyzed")
+
+    mapping.field("xa_avg_grade", "float")
+    mapping.field("xa_histogram_grade", "float")
+    mapping.field("xa_nr_attempts", "integer")
+    mapping.field("xa_nr_views", "integer")
+
+    mapping.save(INDEX_NAME)
+
+    # If we've done this, we always want to re-index all existing
+    # LearningResource instances. This function will probably only
+    # ever be called by migrations.
+    index_resources(LearningResource.objects.all())
+    CONN.indices.refresh()
+
+
+def refresh_index():
+    """
+    Force a refresh instead of waiting for it to happen automatically.
+    This should only be necessary during tests.
+    """
+    CONN.indices.refresh(index=INDEX_NAME)
+
+
+def ensure_vocabulary_mappings(term_info):
+    """
+    Ensure the mapping is properly set in Elasticsearch to always do exact
+    matches on taxonomy terms. Accepts the output of get_resource_terms.
+
+    Calling this function during indexing means that vocabularies do not
+    need to be added to the mapping in advance. This deals with the fact
+    that vocabularies can be added on-the-fly without having to play around
+    with extra signals.
+
+    Args:
+        term_info (dict): Details of terms for a group of LearningResources.
+    """
+    if len(term_info) == 0:
+        return
+
+    # Retrieve current mapping from Elasticsearch.
+    mapping = Mapping.from_es(index=INDEX_NAME, doc_type=DOC_TYPE)
+
+    # Get the field names from the mapping.
+    existing_vocabs = set(mapping.to_dict()["learningresource"]["properties"])
+
+    # Get all the taxonomy names from the data.
+    vocab_names = set()
+    for result in term_info.values():
+        for key in result.keys():
+            vocab_names.add(key)
+    updated = False
+    # Add vocabulary to mapping if necessary.
+    for name in vocab_names:
+        if name in existing_vocabs:
+            continue
+        mapping.field(name, "string", index="not_analyzed", null_value="empty")
```

Comment:
Is `null_value` necessary here since we don't use `null` as a value for vocabs or terms? From the documentation it seems like it's only useful for cases where we want to distinguish between actual `null` and `null` due to missing values.